src/txmempool.cpp

```diff
@@ -1140,3 +1153,47 @@ std::string RemovalReasonToString(const MemPoolRemovalReason& r) noexcept
     }
     assert(false);
 }
+
+std::vector<CTxMemPool::txiter> CTxMemPool::CalculateCluster(const std::vector<uint256>& txids) const
+{
+    AssertLockHeld(cs);
+    std::vector<txiter> cluster{GetIterVec(txids)};
+    if (cluster.size() != txids.size()) {
+        // We can't continue because the caller specified a tx that doesn't exist in the mempool.
+        // Return an empty vector to let them know this failed.
+        return {};
+    }
+    // Reserve total ancestor + descendant counts of each transaction.  This is an approximation; it
+    // may overestimate because transactions may share ancestors/descendants, and may underestimate
+    // because the cluster may include more than just ancestors and descendants.
+    cluster.reserve(std::accumulate(cluster.cbegin(), cluster.cend(), 0, [](size_t sum, const auto it) {
+        return sum + it->GetCountWithAncestors() + it->GetCountWithDescendants() - 1; }));
+    {
+        // Use epoch: visiting an entry means we have added it to the cluster vector. It does not
+        // necessarily mean the entry has been processed.
+        WITH_FRESH_EPOCH(m_epoch);
+        for (const auto& it : cluster) {
+            visited(it);
+        }
+        // i = index of where the list of entries to process starts
+        for (size_t i{0}, to_process_count{txids.size()}; i < to_process_count; ++i) {
+            for (const CTxMemPoolEntry& parent_entry : cluster.at(i)->GetMemPoolParentsConst()) {
+                const auto parent_it = mapTx.iterator_to(parent_entry);
+                if (!visited(parent_it)) {
+                    cluster.push_back(parent_it);
+                    // we still need to process this
+                    ++to_process_count;
```

Comment:
(IIRC we discussed this offline a while ago) We probably want to check here that `to_process_count` doesn't get too large. It's not really feasible to run MiniMiner with a cluster == entire mempool...
Perhaps add a check here so that if we see a cluster size > 1000 (or something), we maybe want to tell the wallet to calculate bump fees in batches, maybe look at fewer UTXOS, or some kind of safe fallback.

Reply:
Marking as resolved as this has been implemented